### PR TITLE
[WASM] Converting runtime JS to TS

### PR DESCRIPTION
**Pull request by Daniel-Genkin**
This PR renames the `dotnet_support.js`, `binding_support.js` and `library_mono.js` files to  `dotnet_support.ts`, `binding_support.ts` and `library_mono.ts` and converts them to use typescript.

## What's changed
Basically everything is as it was only now the files support typescript's type system. So we can better document the functions and have a better development experience as this also has the side effect of adding syntax checking and a few other linting features to the build and VS Code.

Most types can be included directly in the source code, however for some additional types, such as for C functions, there is a new `types` folder. 

Also, now we can use enums to help minimize the guessing with what types are supported for what operations. They can be added to the declaration files via `declare const enum`. During compilation they will be automatically replace with their raw values.

The `tsc` compiler runs during the build but can be invoked manually via `tsc -p <path to runtime folder>` or via `make js-from-ts`  from the MakeFile in the `src/mono/wasm` folder.

## How does the new build work
The same as before except that now, before calling `emcc` to package the js into `dotnet.js`, we call the `tsc` compiler to convert the TS to JS. Then the `emcc` can run as before with the only difference being that the js files are located in a newly generated build folder.

## TODO list
- [x] Convert JS to TS
- [x] Integrate into build
- [x] Add instructions to readme on setting up `tsc` (all you need is to run `npm i -g typescript` to install the tool globally)
- [x] Add `tsc` to CI if it isn't already there (it isn't: https://github.com/dotnet/dotnet-buildtools-prereqs-docker/pull/481)
- [ ] Improve overall code quality
  - [x] Type the parameters and global variables for each file
  - [x] Replace hardcoded switch cases with enums for organization and readability
  - [ ] Get feedback on the types to make sure they are correct and clear

### CODE REVIEW

**Comment by En3Tho** [comment]
Good work. Always nice to see the world becoming a better place by reducing js locs.

**Comment by Daniel-Genkin** [comment]
@terrajobst I am an intern on @lewing 's team. So I am internal to MS.

File: src/mono/wasm/runtime/library_mono.ts [hunk 1]
```diff
@@ -581,8 +540,8 @@ var MonoSupportLib = {
 		mono_wasm_send_dbg_command_with_parms: function (id, command_set, command, command_parameters, length, valtype, newvalue)
```
Comment (thaystg) [inline src/mono/wasm/runtime/library_mono.ts:540]: id, command_set, command are numbers.
command_parameters is a base64 string.
length and valtype are numbers.
newvalue is a string.

**Comment by vargaz** [comment]
Would it be possible to rename these to binding-support.js etc., our convention is to use '-' in file names instead of '_'.

File: src/mono/wasm/runtime/types/debugger-types.d.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        [fn: string]: Function
+    };
+    commands_received:{
+        res_ok: number,
+        res: CommandResult
+    };
+    _call_function_res_cache: object;
+    _next_call_function_res_id: number;
+    _next_id_var: number;
+    var_info: []; // always an empty list - Can it be removed?
```
Comment (thaystg) [inline src/mono/wasm/runtime/types/debugger-types.d.ts:30]: Yes, you can remove.

Comment (Daniel-Genkin) [reply src/mono/wasm/runtime/types/debugger-types.d.ts:30]: Hmm actually turns out you can't it is used in other places. It just isn't referenced directly. I just tried to run the debugger tests with it removed and they failed

**Comment by Daniel-Genkin** [comment]
I think that the Windows WASM build in runtime-stagging is failing because I have not added the tsc compiler to the windows build machines. While I work on that, I am opening up the PR for review so that I can receive feedback about the types and if there are any mistakes. 

## All reviewers please also review the types. I had to guess them so there are bound to be errors or better types available
I specifically left no implicit `any`s and tried to minimize the `unknown` and `any` types. However, for example, if I typed something as a `number` maybe we can make an alias type that is more useful such as `ManagedPointer`.

File: src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props [hunk 3]
```diff
@@ -216,9 +216,9 @@
     <PlatformManifestFileEntry Include="icudt_EFIGS.dat" IsNative="true" />
     <PlatformManifestFileEntry Include="icudt_optimal.dat" IsNative="true" />
     <PlatformManifestFileEntry Include="icudt_optimal_no_CJK.dat" IsNative="true" />
-    <PlatformManifestFileEntry Include="binding_support.js" IsNative="true" />
-    <PlatformManifestFileEntry Include="dotnet_support.js" IsNative="true" />
-    <PlatformManifestFileEntry Include="library_mono.js" IsNative="true" />
+    <PlatformManifestFileEntry Include="binding-support.js" IsNative="true" />
+    <PlatformManifestFileEntry Include="dotnet-support.js" IsNative="true" />
+    <PlatformManifestFileEntry Include="library-mono.js" IsNative="true" />
```
Comment (radical) [inline src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:221]: Why the change from `_` to `-`?

Comment (Daniel-Genkin) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:221]: @vargaz suggested it since the rest of the repo uses this convention

Comment (radical) [reply src/installer/pkg/sfx/Microsoft.NETCore.App/Directory.Build.props:221]: ah, that's correct. I was only looking at the `icudt_*` just above these lines!

**Comment by pavelsavara** [comment]
We have at least 3 different things treated as `number`
- naked pointer
- GCHandle
- JSHandle
could we have separate TS types for those and have them **not** convertible to `number` ?
Perhaps that could be follow-up PR.

File: src/mono/wasm/runtime/binding-support.ts [hunk 4]
```diff
@@ -1491,28 +1482,28 @@ var BindingSupportLib = {
 					//  at a point where we've already disposed of the object the event handler is attached to.
 					// As such, we log here instead of throwing an error. We may want to not log at all...
 					console.log("The delegate target that is being invoked is no longer available.  Please check if it has been prematurely GC'd.");
-					return;
+					return; // TODO make this some number output
 				}
 			}
 
-			var delegateRoot = MONO.mono_wasm_new_root (this.extract_mono_obj (delegate_obj));
+			var delegateRoot = MONO.mono_wasm_new_root (BINDING.extract_mono_obj (delegate_obj));
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/binding-support.ts:1489]: `const`

File: src/mono/wasm/runtime/binding-support.ts [hunk 5]
```diff
@@ -1286,86 +1276,87 @@ var BindingSupportLib = {
...
-			var needs_converter = this._verify_args_for_method_call (args_marshal, args);
+			var needs_converter = BINDING._verify_args_for_method_call (args_marshal, args);
 
 			var buffer = 0, converter = null, argsRootBuffer = null;
 			var is_result_marshaled = true;
 
 			// check if the method signature needs argument mashalling
 			if (needs_converter) {
-				converter = this._compile_converter_for_marshal_string (args_marshal);
+				converter = BINDING._compile_converter_for_marshal_string (args_marshal);
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/binding-support.ts:1286]: `const`

File: src/mono/wasm/runtime/types/binding-types.d.ts [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+// Note that since these are annoated as `declare const enum` they are replaces by tsc with their raw value during compilation
+declare const enum ConverterStepIndirects {
+    UInt32 = "u32",
+    Int32 = "i32",
+    Float = "float",
+    Float64 = "double",
+    Int64 = "i64",
+}
+
+declare const enum CNonPrimativeTypes {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/types/binding-types.d.ts:133]: typo `CNonPrimitiveTypes`

File: src/mono/wasm/runtime/types/binding-types.d.ts [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    DateTimeOffset = 21,
+    Uri = 22,
+    SafeHandle = 23,
+    Int64 = 26, // TODO: Fix this once emscripten offers HEAPI64/HEAPU64 or can return them, currently throws an error
+    UInt64 = 27, // TODO: Fix this once emscripten offers HEAPI64/HEAPU64 or can return them, currently throws an error
+    Char = 29,
+    Undefined = 30
+}
+
+declare const enum CPrimativeTypes {
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/types/binding-types.d.ts:158]: typo

**Comment by pavelsavara** [comment]
It would be good to diff original js *before* and compiled js *after* to make sure that functional changes are reviewed.

**Comment by Daniel-Genkin** [comment]
> We have at least 3 different things treated as `number`
> 
> * naked pointer
> * GCHandle
> * JSHandle
>   could we have separate TS types for those and have them **not** convertible to `number` ?
>   Perhaps that could be follow-up PR.

I am not familiar enough with these details so I agree the best way will probably be to leave them as number for this PR and then over time gradually improve them as we update the files.

**Comment by Daniel-Genkin** [comment]
> It would be good to diff original js _before_ and compiled js _after_ to make sure that functional changes are reviewed.

I did a diff and looks like it is the same code just it uses spaces rather than tabs, some comments are intentionally removed and some spacing is optimized. However this is expected as `tsc` is copying the ts over to the js and then just slightly modifying it to remove the typescript types and I guess also running some basic linter. The underlying code however is not changed.

**Comment by steveisok** [comment]
@lewing @pavelsavara Is there still ongoing work w/ this PR?

**Comment by pavelsavara** [comment]
yes, I will take it over.

**Comment by pavelsavara** [comment]
Replaced by https://github.com/dotnet/runtime/pull/59392
